pkg/cmd/pulumi/destroy.go

```diff
@@ -257,3 +286,52 @@ func newDestroyCmd() *cobra.Command {
 
 	return cmd
 }
+
+// seperateProtected returns a list or unprotected and protected resources
+// respectively. This allows us to safely destroy all resources in the
+// unprotected list without invalidating any resource in the protected list.
+// Protection is contravarient.
+//
+// A
+// B: Parent = A
+// C: Parent = A, Protect = True
+//
+// -->
+//
+// Unprotected: B
+// Protected: A, C
+//
+// We rely on the fact that `resources` is topologically sorted with respect to
+// its dependencies. This function understands that providers live outside this
+// topological sort.
+func seperateProtected(resources []*resource.State) (
+	/*unprotected*/ []*resource.State /*protected*/, []*resource.State) {
+	dg := graph.NewDependencyGraph(resources)
+	transitiveProtected := graph.ResourceSet{}
+	for _, r := range resources {
+		if r.Protect {
+			rProtected := dg.TransitiveDependenciesOf(r)
```

Comment:
So thinking out loud here. Are you sure we're specifying what we want here. The code is traversing TransitiveDependenciesOf and considers them protected, which is the inverse dependency order from DependingOn. So in practice given your own example:

// A
// B: Parent = A
// C: Parent = A, Protect = True
//
// -->
//
// Unprotected: B
// Protected: A, C

We're extending protection to parent of C, so A and C stick around, good.

However we're not extending protection to the children, like this:

// A
// B: Parent = A
// C: Parent = A, Protect = True
// D: Parent = C
//
// -->
//
// Unprotected: B, D
// Protected: A, C

So presumably we'd nuke C but leave D hanging with a deleted parent? is this right?

Actually yeah it seems right. Awesome.